### PR TITLE
[Core] Item plando early locations (and non-early locations)

### PR DESCRIPTION
**Pull request by Alchav**
## What is this fixing or adding?
This is just the early_locations and non_early_locations feature from #1219 split into its own PR. #1219 modifies `get_all_state` and potentially could break things, and needs time to test and ensure stability. Since we are in feature freeze I don't expect (or suggest) we try getting that in at this point. However, I hope this one change can make it in, assuming #1177 is merged.

This also fixes an issue that still exists in #1219 where it only counted the location groups as one location for purposes of counting the number of items that could be placed.

## How was this tested?
Generating LttP with items plando'd into early_locations and non_early_locations. Observed intended effects in spoiler log as well as debugging with a breakpoint to ensure `candidate_locations` only contained the intended locations.

### CODE REVIEW

**Comment by beauxq** [comment]
I see a lot of checking for membership of lists, inside loops. That's slow.

And it looks like we calculate the `early_locations` and `non_early_locations` even if they're not needed?

**Comment by Alchav** [comment]
> I see a lot of checking for membership of lists, inside loops. That's slow.
> 
> And it looks like we calculate the `early_locations` and `non_early_locations` even if they're not needed?

Are you saying I've added checks for list membership in a loop that wasn't there already

File: Fill.py [hunk 1]
```diff
@@ -659,6 +659,16 @@ def failed(warning: str, force: typing.Union[bool, str]) -> None:
...
 
+    swept_state = world.state.copy()
+    swept_state.sweep_for_events()
+    early_locations = {}
+    non_early_locations = {}
+    for player in world.player_ids:
+        reachable_locations = world.get_reachable_locations(swept_state, player)
+        unfilled_locations = world.get_unfilled_locations(player)
+        early_locations[player] = [loc.name for loc in reachable_locations if loc in unfilled_locations]
+        non_early_locations[player] = [loc.name for loc in unfilled_locations if loc not in reachable_locations]
```
Comment (beauxq) [inline Fill.py:670]: Does this look right?
```suggestion
    reachable = frozenset(world.get_reachable_locations(swept_state))
    early_locations: typing.Dict[int, typing.List[str]] = defaultdict(list)
    non_early_locations: typing.Dict[int, typing.List[str]] = defaultdict(list)
    for loc in world.get_unfilled_locations():
        if loc in reachable:
            early_locations[loc.player].append(loc.name)
        else:  # not reachable with swept state
            non_early_locations[loc.player].append(loc.name)
```

Comment (Alchav) [reply Fill.py]: done

**Comment by beauxq** [comment]
> > And it looks like we calculate the `early_locations` and `non_early_locations` even if they're not needed?
> 
> And would you rather I scan every plando block in the multiworld for these group names first

And then, that could be abstracted to a cached function that only runs the first time you find one of the group names.

But I'm not sure whether that's worth it now...

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**